From 3e4b3809e3145daf3c8f0f60e9c1621d44157c83 Mon Sep 17 00:00:00 2001
From: Andrei Gosman <andrei.gosman@gmail.com>
Date: Fri, 2 Oct 2026 15:06:26 +0300
Subject: [PATCH] gtp-u, nfapi: HOST_NAME_MAX, msg_iovlen width and ::bind on
 Darwin

  gtp_itf.h:83:21: error: use of undeclared identifier 'HOST_NAME_MAX'
  gtp_itf.cpp:303:19: error: non-constant-expression cannot be narrowed
  from type 'unsigned int' to 'int' in initializer list
  gtp_itf.cpp:569:49: error: invalid operands to binary expression
  ('__bind<int &, sockaddr *&, unsigned int &>' and 'int')

HOST_NAME_MAX is not POSIX; where it is missing, the POSIX minimum
_POSIX_HOST_NAME_MAX (255, which is also the Darwin limit) serves.
msghdr::msg_iovlen is size_t on glibc and int on Darwin: initialise it
through the field type. gtp_itf.cpp says "using namespace std" and
includes <map>, which brings std::bind in; with libc++ the three-argument
bind() resolves to the template, so name the socket call ::bind. The two
nFAPI sources that call recvmmsg() include darwin_compat.h explicitly,
since they do not reach it through the logging header.
---
 nfapi/oai_integration/socket/socket_vnf.c | 1 +
 nfapi/open-nFAPI/vnf/src/vnf_p7_common.c  | 1 +
 openair3/ocp-gtpu/gtp_itf.cpp             | 4 ++--
 openair3/ocp-gtpu/gtp_itf.h               | 3 +++
 4 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/nfapi/oai_integration/socket/socket_vnf.c b/nfapi/oai_integration/socket/socket_vnf.c
index 15656ca..822bba4 100644
--- a/nfapi/oai_integration/socket/socket_vnf.c
+++ b/nfapi/oai_integration/socket/socket_vnf.c
@@ -4,6 +4,7 @@
 #include "socket_vnf.h"
 
 #include "common/utils/LOG/log.h"
+#include "common/utils/darwin_compat.h"
 #include "nfapi.h"
 #include "vnf/nfapi_nr_vnf.h"
 #include <common/platform_constants.h>
diff --git a/nfapi/open-nFAPI/vnf/src/vnf_p7_common.c b/nfapi/open-nFAPI/vnf/src/vnf_p7_common.c
index 7bf6325..a209015 100644
--- a/nfapi/open-nFAPI/vnf/src/vnf_p7_common.c
+++ b/nfapi/open-nFAPI/vnf/src/vnf_p7_common.c
@@ -15,6 +15,7 @@
 #include <sys/socket.h>
 #include <netinet/in.h>
 #include <arpa/inet.h>
+#include "common/utils/darwin_compat.h"
 
 #include "vnf_p7_common.h"
 
diff --git a/openair3/ocp-gtpu/gtp_itf.cpp b/openair3/ocp-gtpu/gtp_itf.cpp
index 04a1f62..a739e44 100644
--- a/openair3/ocp-gtpu/gtp_itf.cpp
+++ b/openair3/ocp-gtpu/gtp_itf.cpp
@@ -300,7 +300,7 @@ static int gtpv1uCreateAndSendMsg(gtpv1u_bearer_t *bearer,
     .msg_name = to,
     .msg_namelen = sizeof(*to),
     .msg_iov = iov,
-    .msg_iovlen = Msg ? 2U : 1U,
+    .msg_iovlen = static_cast<decltype(msghdr::msg_iovlen)>(Msg ? 2 : 1), // size_t on glibc, int on Darwin
   };
   ssize_t ret = sendmsg(bearer->sock_fd, &m, 0);
   if (ret != (ssize_t) (hdr_len + msgLen)) {
@@ -566,7 +566,7 @@ static int udpServerSocket(openAddr_s addr)
       continue;
     }
 
-    if (bind(sockfd, p->ai_addr, p->ai_addrlen) == -1) {
+    if (::bind(sockfd, p->ai_addr, p->ai_addrlen) == -1) { // not std::bind, see "using namespace std" above
       close(sockfd);
       LOG_W(GTPU, "bind: %s\n", strerror(errno));
       continue;
diff --git a/openair3/ocp-gtpu/gtp_itf.h b/openair3/ocp-gtpu/gtp_itf.h
index 7167b29..74fc473 100644
--- a/openair3/ocp-gtpu/gtp_itf.h
+++ b/openair3/ocp-gtpu/gtp_itf.h
@@ -7,6 +7,9 @@
 
 #include <stdint.h>
 #include <limits.h>
+#ifndef HOST_NAME_MAX /* not in POSIX; Darwin only has the minimum */
+#define HOST_NAME_MAX _POSIX_HOST_NAME_MAX
+#endif
 #include <sys/socket.h>
 #include <netinet/in.h>
 
-- 
2.55.0

